### PR TITLE
Imperial Stockpile 

### PR DESCRIPTION
**Pull request by agrrr3**
This is a WIP implementation of the imperial stockpile based on dilvishs changes to prediction code (Compare #1607 and the original #1229).

I need this to get feedback for the features:
* UI: local/global info in production window
* UI: stockpile info in map window (PP hover)
* UI: waste/stockpile transfer info in map window (PP waste hover)
* game: stockpile with use limit instead of stockpile limit
* game: universal stockpile (non-location dependent) with tech upgrades for stockpile input ratio

The current state of this PR:
There is a stockpile ratio which is like a inverse waste ratio. 0.2 means that 20% of your overproduction goes to stockpile, 80% to waste. Actually I'm thinking of reversing this as "waste ratio" is easier to explain than a "stockpile transfer ratio". Technologies can bring this up to 90%. 
The stockpile is unlimited,  but there is a cap on the amount of resources that can be used from the stockpile per turn (between 2PP and 13PP), enough to notice and for testing; it is expected that we'd be fine tune technologies.

Individual items on the queue by default will NOT draw on the stockpile, they need to be enabled for it via right-click menu. If so enabled, their tooltip will so state, but if not enabled for the stockpile then their tooltip simply does not mention the stockpile. When enabled, there is not currently any specific indicator of just how much PP the item is drawing from the stockpile, but you can easily see that by toggling the stockpile use off and back on to compare.
There is info in production screen how much PP from the stockpile gets drawn in total.

Even without enabling any items to draw on the stockpile, any otherwise wasted PPs will be directed to the stockpile (subject to the waste ratio).

A readout of the current and projected stockpile status is added to the tooltip for Production Points on the MapWnd menu bar. The waste/input ratio is on a tooltip to the waste indicator on the MapWnd menu bar. There is also a new stockpile entry in the MapWnd menu bar using the trade icon and show current stockpile PP and the expected stockpile change (green/red == plus/minus). 

As noted in #1606 the reorganization of the projections code that this relies upon drastically slows down the projections. So far it has still seemed reasonable to me, but for larger empires or with many AIs this may be a more noticeable problem. If you want to check how long projections are taking you can search your freeorion.log (or an AI log) for the phrase ProductionQueue::Update: Projections took

The new technologies are:
PRO_IMPERIAL_STOCKPILE_1   (you get that from the start)
Predictive Stockpiling

PRO_IMPERIAL_STOCKPILE_2
Generic Supplies

PRO_IMPERIAL_STOCKPILE_3
Interstellar Entanglement Factory

CON_TRANS_DESIGN
Transcendent Design

### CODE REVIEW

File: Empire/Empire.cpp [hunk 1]
```diff
@@ -107,6 +112,44 @@ namespace {
         return retval;
     }
 
+    float CalculateNewStockpile(int empire_id, int stockpile_location_id, float starting_stockpile,
+                                         const std::map<std::set<int>, float>& available_pp,
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: fix indenting

File: Empire/Empire.cpp [hunk 2]
```diff
@@ -256,9 +292,9 @@ namespace {
                 continue;
             }
 
-
             // get max contribution per turn and turns to build at max contribution rate
-            int location_id = (queue_element.item.CostIsProductionLocationInvariant() ? INVALID_OBJECT_ID : queue_element.location);
+            int location_id = (queue_element.item.CostIsProductionLocationInvariant() ?
+                        INVALID_OBJECT_ID : queue_element.location);
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: indent 4 spaces, not 12

File: Empire/Empire.cpp [hunk 3]
```diff
@@ -988,11 +1081,24 @@ void ProductionQueue::Update() {
...
+                                local_stockpile_use_rate, remote_stockpile_use_rate,
+                                queue_element_groups,
                                 queue_item_costs_and_times, is_producible, m_queue,
-                                m_object_group_allocated_pp, m_projects_in_progress);
+                                m_object_group_allocated_pp, m_object_group_allocated_stockpile_pp,
+                                m_projects_in_progress, false);
+
+    //update expected new stockpile amount
+    m_expected_new_stockpile_amount = CalculateNewStockpile(m_empire_id, stockpile_location_id, pp_in_stockpile, available_pp, 
+                                                       m_object_group_allocated_pp, m_object_group_allocated_stockpile_pp);
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: indent to match previous

File: Empire/Empire.cpp [hunk 4]
```diff
@@ -1037,147 +1133,72 @@ void ProductionQueue::Update() {
...
+    sim_time_start = boost::posix_time::ptime(boost::posix_time::microsec_clock::local_time()); 
+    std::map<std::set<int>, float>  allocated_pp;
+    float sim_available_stockpile = available_stockpile;
+    float sim_pp_in_stockpile = available_stockpile;
+    std::map<std::set<int>, float>  allocated_stockpile_pp;
+    int dummy_int = 0;
+    
+    for (int sim_turn = 1; sim_turn <= TOO_MANY_TURNS; sim_turn ++) {
+        if ((boost::posix_time::ptime(boost::posix_time::microsec_clock::local_time()) -
+                    sim_time_start).total_microseconds()*1e-6 >= TOO_LONG_TIME)
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: weird indenting

File: Empire/ResourcePool.cpp [hunk 5]
```diff
@@ -28,13 +28,15 @@ int ResourcePool::StockpileObjectID() const
...
 
-float ResourcePool::Output() const {
+float ResourcePool::TotalOutput() const {
     float retval = 0.0f;
     for (const auto& entry : m_connected_object_groups_resource_output)
     { retval += entry.second; }
     return retval;
 }
 
+std::map<std::set<int>, float> ResourcePool::Output() const { return  m_connected_object_groups_resource_output; }
```
Comment (geoffthemedio) [inline Empire/ResourcePool.cpp]: weird double space after `return`

File: default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt [hunk 6]
```diff
@@ -0,0 +1,94 @@
+BuildingType
+    name = "BLD_INTERSPECIES_ACADEMY"
```
Comment (geoffthemedio) [inline default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt:2]: is this relevant to this pull request?

File: default/scripting/techs/specials/TESTING_UNLOCKS.focs.txt [hunk 7]
```diff
@@ -0,0 +1,16 @@
+Tech
+    name = "TEST_TECH_UNLOCK_1"
```
Comment (geoffthemedio) [inline default/scripting/techs/specials/TESTING_UNLOCKS.focs.txt]: ?

File: default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt (hunk 6 above)
Comment (agrrr3) [reply default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt:2]: yes

File: Empire/Empire.cpp [hunk 8]
```diff
@@ -1129,8 +1129,8 @@ void ProductionQueue::Update() {
     int dummy_int = 0;
     
     for (int sim_turn = 1; sim_turn <= TOO_MANY_TURNS; sim_turn ++) {
-        if ((boost::posix_time::ptime(boost::posix_time::microsec_clock::local_time()) -
-                    sim_time_start).total_microseconds()*1e-6 >= TOO_LONG_TIME)
+        long sim_time_until_now =  (boost::posix_time::ptime(boost::posix_time::microsec_clock::local_time()) - sim_time_start).total_microseconds();
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: extra space after `=`

File: default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                Building name = "BLD_INTERSPECIES_ACADEMY"
+                OwnedBy empire = Source.Owner
+            ]
+        ]
+        Number low = 0 high = 0 condition = And [
+            Building name = "BLD_INTERSPECIES_ACADEMY"
+            Species name = RootCandidate.Species
+            OwnedBy empire = RootCandidate.Owner
+        ]
+        Number low = 0 high = 0 condition = Described description = "Academies of this species already exist in the empire" condition = And [
```
Comment (geoffthemedio) [inline default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt]: Use stringtable references, not hard-coded english in FOCS.

File: default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                    NoStringtableLookup
+                    icon = "icons/building/blackhole.png"
+                    parameters = [
+                        tag = "planet" data = Source.PlanetID
+                        tag = "species" data = Source.Planet.Species
+                        tag = "rawtext" data = EmpireMeterValue empire = Source.Owner meter = "METER_IMPERIAL_PP_TRANSFER_EFFICIENCY"
+                    ]
+                    empire = Source.Owner
+                SetEmpireMeter empire = Source.Owner meter = "METER_IMPERIAL_PP_TRANSFER_EFFICIENCY" value = Value + 0.05
+                SetEmpireMeter empire = Source.Owner meter = "METER_IMPERIAL_PP_USE_LIMIT"           value = Value + 1
```
Comment (geoffthemedio) [inline default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt]: needs limits to ensure efficiency doesn't go over 1

Comment (agrrr3) [reply default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt]: Also I cap the efficiency at 1.0f in the backend see commit a5faf12433798f4faa945cd5608529da39380867 .

File: default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt (hunk 9 above)
Comment (agrrr3) [reply default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt]: If i remember correctly this is to give a better description why you cant build the academy.
Does described work with stringtables? Or is there another way to add a description of that condition?

Comment (dbenage-cx) [reply default/scripting/buildings/INTERSPECIES_ACADEMY.focs.txt]: Not looked at it closely, the only current use is a stringtable lookup: [CULTURE_ARCHIVES.focs.txt:7](https://github.com/freeorion/freeorion/blob/master/default/scripting/buildings/CULTURE_ARCHIVES.focs.txt#L7)

**Comment by agrrr3** [comment]
think the branch is in a state where it is actually useable
have some issues with the ui
but somebody should have a look at whats good and whats not

**Comment by agrrr3** [comment]
only thing missing imho is a frontier skill for species which give a bonus to stockpile meters
but i would like to add that later

**Comment by dbenage-cx** [comment]
Curious how this PR relates to others with similar names.
My guess is this PR is blocked by #1607, which in turn supersedes #1229?

**Comment by agrrr3** [comment]
I forked this branch off of #1607 (hence the branch name) and i forgot about #1229 which was my old code where i developed the stockpile until i hit the road block in the production prediction code (and dilvish was taking hints and code from there).
Both are obsolete. I closed mine now.

**Comment by geoffthemedio** [comment]
In the production info box, I'm not sure what the extra numbers are for, and there is some odd extra duplicated text that needs cleaning up.

extra_text

**Comment by geoffthemedio** [comment]
The amount indicating what's available at the selected planet is wrong. See attached. Those 11.4 PP are in a different resource sharing group.

available

**Comment by geoffthemedio** [comment]
The tooltip for the production resource icon should have a line indicating how much is being sent to the stockpile from unused production points. Better if it also indicated the transfer efficiency factor to explain why that's how much is being transferred. That info shouldn't just be in the (!) warning icon.

stockpile_tooltip

The production output and production stockpile icons would probably be better grouped closer together as well. Could be done as part of reworking how all those icons are spaced out, as they can look rather awkward when the window is rather wide.

File: default/stringtables/en.txt [hunk 11]
```diff
@@ -9899,6 +9985,50 @@ While transforming an entire population into beings of pure energy would be impr
...
+IMPERIAL_STOCKPILE_2_DESC
+'''By carefully choosing supply chains, materials, and designs, Generic Supplies are created which can easily be decomposited and reused for most industrial products.
+
+Generic Supplies can be created a long time before the demand occurs. Supply packages usually include bootstrap devices which reconstruct the facilities for reconstructing new products from the Generic Supplies.
+Reaching this state of sophistication, Generic Supplies can be used to construct new technology products which were not available at the time the supplies were produced.
+
+This frees the imperial stockpile service from having to know which goods will be demanded exactly. Predicting the location is enough.
+
+This is an upgrade technology for the imperial stockpile.
+It adds 20% to the [[metertype METER_IMPERIAL_PP_TRANSFER_EFFICIENCY]] and allows to draw 3PP/turn extra from the stockpile (See [[metertype METER_IMPERIAL_PP_USE_LIMIT]]).'''
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: "allows to draw" is grammatically wrong; use "allows drawing" or "allows the withdrawl of up to"

put spaces between `3` and `PP` and around `/`

File: universe/Enums.h [hunk 12]
```diff
@@ -102,7 +102,7 @@ GG_ENUM(MeterType,
     METER_MAX_TROOPS,
 
     METER_POPULATION,
-    METER_INDUSTRY,
+    METER_INDUSTRY,
```
Comment (geoffthemedio) [inline universe/Enums.h]: ?

File: universe/Enums.h [hunk 13]
```diff
@@ -124,7 +124,10 @@ GG_ENUM(MeterType,
     METER_DETECTION,
     METER_SPEED,
 
-    NUM_METER_TYPES
+    NUM_METER_TYPES,
+
+    METER_IMPERIAL_PP_USE_LIMIT,
```
Comment (geoffthemedio) [inline universe/Enums.h]: keep `NUM_METER_TYPES` last

File: util/Order.cpp [hunk 14]
```diff
@@ -980,6 +986,14 @@ void ProductionQueueOrder::ExecuteImpl() const {
                 DebugLogger() << "ProductionQueueOrder: unpausing production";
                 empire->ResumeProduction(m_index);
 
+            } else if (m_use_imperial_pp == USE_IMPERIAL_PP) {
+                DebugLogger() << "ProductionQueueOrder: allow use of imperial PP stockpile";
+                empire->AllowUseImperialPP(m_index, true);; //
```
Comment (geoffthemedio) [inline util/Order.cpp]: double `;`
pointless `//`

File: util/SerializeEmpire.cpp [hunk 15]
```diff
@@ -69,7 +69,8 @@ void ProductionQueue::Element::serialize(Archive& ar, const unsigned int version
         & BOOST_SERIALIZATION_NVP(turns_left_to_next_item)
         & BOOST_SERIALIZATION_NVP(turns_left_to_completion)
         & BOOST_SERIALIZATION_NVP(rally_point_id)
-        & BOOST_SERIALIZATION_NVP(paused);
+        & BOOST_SERIALIZATION_NVP(paused)
+        & BOOST_SERIALIZATION_NVP(allowed_imperial_stockpile_use);
```
Comment (geoffthemedio) [inline util/SerializeEmpire.cpp:74]: use `BOOST_CLASS_VERSION` to preserve backwards save compatibility.

File: default/stringtables/en.txt [hunk 16]
```diff
@@ -9990,6 +10120,18 @@ Neutronium Extraction
...
 The core of a [[STAR_NEUTRON]] star consists of degenerate neutron matter, compressed by gravity to a density far greater than the heaviest normal matter. This exotic material has numerous applications once extracted from the star, allowing production of otherwise impossible structures and ship parts.
 
+PRO_IMPERIAL_SUPPLY_1
+Imperial Supply Stockpiling
+
+IMPERIAL_STOCKPILE_SHORT_DESC
+Allows for a small Imperial Production Stockpile
+
+IMPERIAL_STOCKPILE_1_DESC
+'''Unused PP, which would otherwise be wasted, can have a portion instead be retained and transferred to the Imperial Production Stockpile.  An item on the production queue which does not have its full turn allotment of PP available to it via the normal supply mechanism, can draw on the Imperial Stockpile if the item has been explicitly enabled for doing so (via right-click menu on the Production Queue).
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: overly verbose text. simplify.

File: default/stringtables/en.txt [hunk 17]
```diff
@@ -9899,6 +9985,50 @@ While transforming an entire population into beings of pure energy would be impr
...
+
+This is an upgrade technology for the imperial stockpile.
+It adds 20% to the [[metertype METER_IMPERIAL_PP_TRANSFER_EFFICIENCY]] and allows to draw 3PP/turn extra from the stockpile (See [[metertype METER_IMPERIAL_PP_USE_LIMIT]]).'''
+
+PRO_IMPERIAL_STOCKPILE_3
+Interstellar Entanglement Factory
+
+IMPERIAL_STOCKPILE_3_DESC
+'''Quantum entanglement of stockpiled goods allows specialists to operate machines remotely from other star systems.
+Putting the empires best engineers directly in charge of building using the stockpiled goods helps making the most out of these.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: empire's

whole paragraph of flavour text isn't great

File: default/stringtables/en.txt [hunk 18]
```diff
@@ -2851,10 +2851,16 @@ MAP_PROD_WASTED_TITLE
...
-# %2% number of production points currently not used for production.
+# %2% number of production points currently neither used for production nor for stockpile.
+# %3% yield ratio of importing PP into the imperial stockpile (the other part goes to waste)
+# %4% number of production points transferred to the imperial stockpile.
+# %5% number of production points currently neither used for production nor for stockpile.
 MAP_PROD_WASTED_TEXT
 '''Total Production : %1%
-Wasted Production : %2%
+Overproduction : %2%
+Stockpile Import Ratio: %3%
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: "Import" unclear

File: default/stringtables/en.txt [hunk 19]
```diff
@@ -2851,10 +2851,16 @@ MAP_PROD_WASTED_TITLE
...
 # %1% number of production points generated.
-# %2% number of production points currently not used for production.
+# %2% number of production points currently neither used for production nor for stockpile.
+# %3% yield ratio of importing PP into the imperial stockpile (the other part goes to waste)
+# %4% number of production points transferred to the imperial stockpile.
+# %5% number of production points currently neither used for production nor for stockpile.
 MAP_PROD_WASTED_TEXT
 '''Total Production : %1%
-Wasted Production : %2%
+Overproduction : %2%
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: Over -> Excess ?

File: UI/ResourceBrowseWnd.h [hunk 20]
```diff
@@ -32,7 +34,17 @@ class ResourceBrowseWnd : public GG::BrowseInfoWnd {
     std::shared_ptr<GG::Label>  m_target_points_label;
     std::shared_ptr<GG::Label>  m_target_points;
     std::shared_ptr<GG::Label>  m_target_points_P_label;
-
+    bool        m_show_stockpile;
```
Comment (geoffthemedio) [inline UI/ResourceBrowseWnd.h]: indent to match

File: UI/ResourceBrowseWnd.h [hunk 21]
```diff
@@ -23,6 +24,7 @@ class ResourceBrowseWnd : public GG::BrowseInfoWnd {
     GG::GL2DVertexBuffer    m_buffer;
```
Comment (geoffthemedio) [inline UI/ResourceBrowseWnd.h]: align indenting

File: Empire/Empire.cpp [hunk 22]
```diff
@@ -206,9 +243,11 @@ namespace {
 
         projects_in_progress = 0;
         allocated_pp.clear();
-
+        allocated_stockpile_pp.clear();
+        float dummy_pp_source = 0;
+        
         //DebugLogger() << "queue size: " << queue.size();
-
+
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: ?

**Comment by agrrr3** [comment]
Looks like you cant answer directly to comments on github(?). Anyway this goes @geoffthemedio  UI comments. Im happy somebody had finally a look at it :) 

Regarding the first two comments:
With all that extra information the production info box took too much screen space on my machine.
So I condensed the empire and local production info into two columns.
The left column is empire production and the right column is local production. To the far left are the global labels. To the far right (partially visible) are the local labels.
My suggestion would be to add Column headers, and get rid of local labels.

Regarding the third comment:
The tooltip for the production resource icon should or shouldnt include stockpile information at all. 
Did you see the trade icon which i reused for stockpile resource icon? Please compare with that.

File: default/stringtables/en.txt (hunk 19 above)
Comment (agrrr3) [reply default/stringtables/en.txt]: I'd think overproduction would be correct (technical?) term. I don't know about daily usage though.
wikipedia: In economics, overproduction, oversupply, excess of supply or glut refers to excess of supply over demand of products being offered to the market.
Also google search for overproduction results in 2M hits and "excess production" is 400k

File: default/stringtables/en.txt (hunk 18 above)
Comment (agrrr3) [reply default/stringtables/en.txt]: Yeah. I know. Will start a poll on the forums if i should go inverse and call it Waste Ratio

File: default/stringtables/en.txt (hunk 17 above)
Comment (agrrr3) [reply default/stringtables/en.txt]: You mean the idea is not great or the writing? Or both?

Comment (agrrr3) [reply default/stringtables/en.txt]: The idea is that you build a control room and its clone out of entangled particles then you ship one to the remote factory location and hook it up there. What happens to one happens to the other ;)

File: default/stringtables/en.txt (hunk 19 above)
Comment (geoffthemedio) [reply default/stringtables/en.txt]: Overproduction is having too much of something produced, not having too much capacity to produce something. The latter is being described. Excess production could also mean the latter, but we're referring to the capacity, or specifically the amount of production points, which is not an existing economics concept, but which is well-described by "excess".

File: default/stringtables/en.txt (hunk 17 above)
Comment (geoffthemedio) [reply default/stringtables/en.txt]: References to "specialists" or "empire[']s best engineers" not good style

**Comment by geoffthemedio** [comment]
I thought I made a comment, but I don't see it: Need something other than the trade icon to indicate the production stockpile.

The two columns don't work now due to layout issues, and because the title bar says "At Plane Name", but then the first (or really only) thing you see is the empire totals.

File: default/stringtables/en.txt (hunk 19 above)
Comment (agrrr3) [reply default/stringtables/en.txt]: So "Total Production" should be "Production Capacity" and "Overproduction" should be "Excess Production" (or maybe "Demand Deficit")?

Comment (agrrr3) [reply default/stringtables/en.txt]: Or rather "Unused Capacity"?

**Comment by agrrr3** [comment]
@geoffthemedio I couldnt think of a better icon than the trade icon (it captures the in/out semantics of the stockpile quite well) and we dont (plan) to use it, do we? 

Who of the contributors could help me with the icon (and naming and the UI in general)? Moving pixels is not my forté.

**Comment by agrrr3** [comment]
JFYI Started a UI discussion (SP vs PP) on the forums
http://www.freeorion.org/forum/viewtopic.php?f=6&t=10703

update: ...will stick with PP if no problems arise

File: util/SerializeEmpire.cpp (hunk 15 above)
Comment (agrrr3) [reply util/SerializeEmpire.cpp:74]: Not sure how I do that. As I couldnt find any BOOST_CLASS_VERSION for Empire or Production::Element.
Should I just add a BOOST_CLASS_VERSION(Production::Element, 1) in SerializeEmpire?

**Review by dbenage-cx — COMMENTED** [review]
Nice progress so far.

Is (e.g.) `METER_IMPERIAL_PP_TRANSFER_EFFICIENCY` intended for other use, or only with relation to stockpile?

One point of confusion for me is the duplicate info in production icon and new stockpile icon.
Do these values differ somehow?  If not, might add the current stockpile value next to production (and if its supported, an up/down arrow to indicate changes)

File: Empire/Empire.cpp [hunk 23]
```diff
@@ -107,6 +112,36 @@ namespace {
...
+        float new_contributions = 0.0f;
+        for (auto const& available_group: available_pp) {
+            auto alloc_it = allocated_pp.find(available_group.first);
+            float allocated_here = (alloc_it == allocated_pp.end()) ? 0.0f : alloc_it->second;
+            float excess_here = available_group.second - allocated_here;
+            if (excess_here < EPSILON)
+                continue;
+            // Transfer excess to stockpile
+            new_contributions += excess_here * stockpile_transfer_efficiency;
+            DebugLogger() << "allocated_here " << allocated_here;
```
Comment (dbenage-cx) [inline Empire/Empire.cpp:136]: If retained in final form, please change to `TraceLogger`, combine uses where possible, and provide the iteration key.
In test run, I have 2000+ log lines for this function and uses on ln 330-332, within a single turn.

File: util/SerializeEmpire.cpp (hunk 15 above)
Comment (geoffthemedio) [reply util/SerializeEmpire.cpp:74]: Not in SerializeEmpire, but otherwise yes. See other uses of it around the code.

**Comment by geoffthemedio** [comment]
Can you fix the Unused Production tooltip layout to be like the others, with neat columns?
stockpile_tooltips

**Comment by agrrr3** [comment]
Well, I am not sure.
All the resource tooltips the same base windows and use a (rather
cumbersome to code) column layout.

The waste tooltips (also for research) use some other base class with
simple text contents.

I'd hope there is some way to display tabular data in the UI framework. If
it exists it is not used there.

2017-09-25 23:32 GMT+02:00 Geoff <notifications@github.com>:

**Comment by geoffthemedio** [comment]
Just copy what the resource tooltips do...?

**Comment by agrrr3** [comment]
Not sure if I should files ProductionWastedBrowseWnd.h and .cpp

Also FleetDetailBrowseWnd in MapWnd uses similar but different layout. I
think all the Browse windows in the Map top should use the same
style/layout. But which one should I use now?

2017-09-26 15:58 GMT+02:00 Geoff <notifications@github.com>:

**Comment by geoffthemedio** [comment]
Whichever you prefer. If you base it on FleetDetailBrowseWnd, you'll probably want/need to rework the layout to have the units column. The ResourceBrowseWnd at least uses buffers to render instead of just immediate GL calls.

**Comment by geoffthemedio** [comment]
Also now needs a rebase.

**Comment by geoffthemedio** [comment]
@agrrr3 Will you get back to this soon?

**Comment by agrrr3** [comment]
I'm swamped for real atm and I'm not fast with gigi.
I don't think i can contribute anything meaningful in the next 2 weeks.
Maybe we can merge and somebody fixes the UI?

Am 19.10.2017 10:22 schrieb "Geoff" <notifications@github.com>:

**Comment by Vezzra** [comment]
> Maybe we can merge and somebody fixes the UI?

Better if that somebody (whoever wants to take on this) pushes these fixes to this PR. The implementation should be basically complete and all obvious issues addressed before merge IMO.

**Comment by Vezzra** [comment]
Superseded by #1829.
